File: app/src/main/java/com/weiqi/ui/screens/SetupScreen.kt
package com.weiqi.ui.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.FilterChip
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.weiqi.ai.AiDifficulty
import com.weiqi.engine.GameConfig
import com.weiqi.engine.StoneColor
import com.weiqi.ui.components.ZenCard

data class GameSetup(
    val config: GameConfig,
    val opponent: Opponent,
    val aiColor: StoneColor,
    val aiDifficulty: AiDifficulty
)

@Composable
fun SetupScreen(
    isAi: Boolean,
    onStart: (GameSetup) -> Unit
) {
    var size by remember { mutableIntStateOf(9) }
    var komi by remember { mutableStateOf(7.5) }
    var handicap by remember { mutableIntStateOf(0) }
    var aiColor by remember { mutableStateOf(StoneColor.WHITE) }
    var aiDifficulty by remember { mutableStateOf(AiDifficulty.BEGINNER) }

    Column(
        Modifier
            .fillMaxSize()
            .background(MaterialTheme.colorScheme.background)
            .verticalScroll(rememberScrollState())
            .padding(horizontal = 20.dp, vertical = 16.dp),
        verticalArrangement = Arrangement.spacedBy(12.dp)
    ) {
        Column {
            Text(if (isAi) "Vs. AI" else "Local Match",
                style = MaterialTheme.typography.labelMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant)
            Text("Game setup",
                style = MaterialTheme.typography.headlineMedium,
                fontWeight = FontWeight.SemiBold)
        }

        ZenCard(modifier = Modifier.fillMaxWidth()) {
            Column(
                modifier = Modifier.fillMaxSize(),
                verticalArrangement = Arrangement.spacedBy(20.dp)
            ) {
                ChipSection(
                    title = "BOARD SIZE",
                    options = listOf(9, 13, 19),
                    isSelected = { it == size },
                    label = { "${it}×${it}" },
                    onSelect = { size = it }
                )
                ChipSection(
                    title = "KOMI",
                    options = listOf(0.5, 5.5, 6.5, 7.5),
                    isSelected = { it == komi },
                    label = { it.toString() },
                    onSelect = { komi = it }
                )
                Column {
                    SectionLabel("HANDICAP")
                    Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
                        listOf(
                            listOf(0, 2, 3),
                            listOf(4, 5, 6),
                            listOf(7, 8, 9)
                        ).forEach { rowOpts ->
                            Row(
                                modifier = Modifier.fillMaxWidth(),
                                horizontalArrangement = Arrangement.spacedBy(8.dp)
                            ) {
                                rowOpts.forEach { h ->
                                    FilterChip(
                                        selected = handicap == h,
                                        onClick = { handicap = h },
                                        label = { Text("$h") },
                                        shape = RoundedCornerShape(14.dp),
                                        modifier = Modifier.weight(1f)
                                    )
                                }
                            }
                        }
                    }
                }
                if (isAi) {
                    Column {
                        SectionLabel("AI PLAYS")
                        Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                            listOf(StoneColor.BLACK, StoneColor.WHITE).forEach { c ->
                                FilterChip(
                                    selected = aiColor == c,
                                    onClick = { aiColor = c },
                                    label = { Text(c.name) },
                                    shape = RoundedCornerShape(14.dp)
                                )
                            }
                        }
                    }
                    Column {
                        SectionLabel("AI DIFFICULTY")
                        Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                            AiDifficulty.entries.forEach { d ->
                                FilterChip(
                                    selected = aiDifficulty == d,
                                    onClick = { aiDifficulty = d },
                                    label = { Text(d.label) },
                                    shape = RoundedCornerShape(14.dp)
                                )
                            }
                        }
                        Text(
                            aiDifficulty.description,
                            style = MaterialTheme.typography.labelSmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                            modifier = Modifier.padding(top = 6.dp)
                        )
                    }
                }
            }
        }

        Button(
            onClick = {
                onStart(
                    GameSetup(
                        config = GameConfig(boardSize = size, komi = komi, handicap = handicap),
                        opponent = if (isAi) Opponent.AI else Opponent.HUMAN,
                        aiColor = aiColor,
                        aiDifficulty = aiDifficulty
                    )
                )
            },
            colors = ButtonDefaults.buttonColors(
                containerColor = MaterialTheme.colorScheme.primary,
                contentColor = MaterialTheme.colorScheme.onPrimary
            ),
            shape = RoundedCornerShape(16.dp),
            modifier = Modifier
                .fillMaxWidth()
                .height(60.dp)
        ) {
            Text("BEGIN GAME", style = MaterialTheme.typography.labelLarge)
        }
    }
}

@Composable
private fun <T> ChipSection(
    title: String,
    options: List<T>,
    isSelected: (T) -> Boolean,
    label: (T) -> String,
    onSelect: (T) -> Unit
) {
    Column {
        SectionLabel(title)
        Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
            options.forEach { opt ->
                FilterChip(
                    selected = isSelected(opt),
                    onClick = { onSelect(opt) },
                    label = { Text(label(opt)) },
                    shape = RoundedCornerShape(14.dp)
                )
            }
        }
    }
}

@Composable
private fun SectionLabel(text: String) {
    Text(
        text,
        style = MaterialTheme.typography.labelSmall,
        color = MaterialTheme.colorScheme.onSurfaceVariant,
        modifier = Modifier.padding(bottom = 8.dp)
    )
}

File: lib/src/ui/screens/setup_screen.dart
import 'package:flutter/material.dart';

import '../../domain/bots/bot_catalog.dart';
import '../../domain/bots/bot_profile.dart';
import '../../domain/clock/time_control.dart';
import '../../domain/models.dart';
import '../components/zen_components.dart';
import 'bot_picker_screen.dart';
import 'game_view_model.dart';

class GameSetup {
  final GameConfig config;
  final Opponent opponent;
  final StoneColor aiColor;
  final AiDifficulty aiDifficulty;
  final TimeControl timeControl;
  final BotProfile? bot;
  const GameSetup({
    required this.config,
    required this.opponent,
    required this.aiColor,
    required this.aiDifficulty,
    required this.timeControl,
    this.bot,
  });
}

enum _TimePreset { noClock, blitz, rapid, fischer5p3, byoYomi, canadian }

extension _TimePresetX on _TimePreset {
  String get label => switch (this) {
        _TimePreset.noClock => 'None',
        _TimePreset.blitz => '5 min',
        _TimePreset.rapid => '15 min',
        _TimePreset.fischer5p3 => '5 + 3',
        _TimePreset.byoYomi => 'Byo-yomi',
        _TimePreset.canadian => 'Canadian',
      };

  TimeControl get control => switch (this) {
        _TimePreset.noClock => const TimeControl.none(),
        _TimePreset.blitz => const TimeControl.absolute(mainSeconds: 5 * 60),
        _TimePreset.rapid => const TimeControl.absolute(mainSeconds: 15 * 60),
        _TimePreset.fischer5p3 =>
          const TimeControl.fischer(mainSeconds: 5 * 60, incrementSeconds: 3),
        _TimePreset.byoYomi => const TimeControl.byoYomi(
            mainSeconds: 10 * 60, periods: 3, periodSeconds: 30),
        _TimePreset.canadian => const TimeControl.canadian(
            mainSeconds: 10 * 60, stonesPerPeriod: 20, periodSeconds: 5 * 60),
      };
}

class SetupScreen extends StatefulWidget {
  final bool isAi;
  final ValueChanged<GameSetup> onStart;

  const SetupScreen({super.key, required this.isAi, required this.onStart});

  @override
  State<SetupScreen> createState() => _SetupScreenState();
}

class _SetupScreenState extends State<SetupScreen> {
  int size = 9;
  Ruleset ruleset = Ruleset.chinese;
  double komi = 7.5;
  int handicap = 0;
  StoneColor aiColor = StoneColor.white;
  AiDifficulty aiDifficulty = AiDifficulty.beginner;
  _TimePreset timePreset = _TimePreset.rapid;
  BotProfile? selectedBot = BotCatalog.byId('kiri');
  GameVariant variant = GameVariant.standard;

  void _onRulesetChanged(Ruleset r) {
    final defaults = RulesetDefaults.of(r);
    setState(() {
      ruleset = r;
      komi = defaults.komi;
    });
  }

  @override
  Widget build(BuildContext context) {
    final scheme = Theme.of(context).colorScheme;
    final text = Theme.of(context).textTheme;
    final defaults = RulesetDefaults.of(ruleset);
    return SingleChildScrollView(
      padding: const EdgeInsets.symmetric(horizontal: 20, vertical: 16),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          Text(widget.isAi ? 'Practice Match' : 'Local Match',
              style:
                  text.labelMedium?.copyWith(color: scheme.onSurfaceVariant)),
          Text('Game setup',
              style:
                  text.headlineMedium?.copyWith(fontWeight: FontWeight.w600)),
          const SizedBox(height: 12),
          ZenCard(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.stretch,
              children: [
                _RulesetSection(
                  selected: ruleset,
                  onSelect: _onRulesetChanged,
                ),
                const SizedBox(height: 20),
                _ChipSection<GameVariant>(
                  title: 'GAME MODE',
                  options: GameVariant.values,
                  selected: variant,
                  label: (v) => v.label,
                  onSelect: (v) => setState(() => variant = v),
                ),
                const SizedBox(height: 20),
                _ChipSection<int>(
                  title: 'BOARD SIZE',
                  options: const [9, 13, 19],
                  selected: size,
                  label: (v) => '$v×$v',
                  onSelect: (v) => setState(() => size = v),
                ),
                const SizedBox(height: 20),
                _ChipSection<double>(
                  title: 'KOMI',
                  options: const [0.5, 5.5, 6.5, 7.0, 7.5, 8.0],
                  selected: komi,
                  label: (v) => v == v.roundToDouble()
                      ? v.toStringAsFixed(0)
                      : v.toString(),
                  onSelect: (v) => setState(() => komi = v),
                ),
                const SizedBox(height: 20),
                const _SectionLabel('HANDICAP'),
                const SizedBox(height: 8),
                ..._handicapRows(),
                const SizedBox(height: 20),
                const _SectionLabel('TIME CONTROL'),
                const SizedBox(height: 8),
                SizedBox(
                  height: 44,
                  child: ListView.separated(
                    scrollDirection: Axis.horizontal,
                    itemCount: _TimePreset.values.length,
                    separatorBuilder: (_, __) => const SizedBox(width: 8),
                    itemBuilder: (context, i) {
                      final preset = _TimePreset.values[i];
                      return SizedBox(
                        width: 110,
                        child: ZenOptionButton(
                          label: preset.label,
                          selected: preset == timePreset,
                          onTap: () => setState(() => timePreset = preset),
                        ),
                      );
                    },
                  ),
                ),
                const SizedBox(height: 6),
                Text(
                  timePreset.control.describe(),
                  style: text.labelSmall
                      ?.copyWith(color: scheme.onSurfaceVariant),
                ),
                if (widget.isAi) ...[
                  const SizedBox(height: 20),
                  const _SectionLabel('OPPONENT'),
                  const SizedBox(height: 8),
                  _BotRow(
                    bot: selectedBot,
                    onTap: _pickBot,
                  ),
                  const SizedBox(height: 20),
                  const _SectionLabel('OPPONENT PLAYS'),
                  const SizedBox(height: 8),
                  Row(
                    children: [
                      Expanded(
                        child: _optionButton(
                            'BLACK',
                            aiColor == StoneColor.black,
                            () => setState(() => aiColor = StoneColor.black)),
                      ),
                      const SizedBox(width: 8),
                      Expanded(
                        child: _optionButton(
                            'WHITE',
                            aiColor == StoneColor.white,
                            () => setState(() => aiColor = StoneColor.white)),
                      ),
                    ],
                  ),
                ],
              ],
            ),
          ),
          const SizedBox(height: 12),
          SizedBox(
            height: 60,
            child: FilledButton(
              onPressed: () => widget.onStart(GameSetup(
                config: GameConfig(
                  boardSize: size,
                  ruleset: ruleset,
                  komi: komi,
                  handicap: handicap,
                  allowSuicide: defaults.allowSuicide,
                  superkoMode: defaults.superkoMode,
                  variant: variant,
                ),
                opponent: widget.isAi ? Opponent.ai : Opponent.human,
                aiColor: aiColor,
                aiDifficulty: widget.isAi
                    ? (selectedBot?.engine ?? aiDifficulty)
                    : aiDifficulty,
                timeControl: timePreset.control,
                bot: widget.isAi ? selectedBot : null,
              )),
              child: Text('BEGIN GAME',
                  style: text.labelLarge?.copyWith(color: scheme.onPrimary)),
            ),
          ),
        ],
      ),
    );
  }

  List<Widget> _handicapRows() {
    const rows = <List<int>>[
      [0, 2, 3],
      [4, 5, 6],
      [7, 8, 9],
    ];
    final widgets = <Widget>[];
    for (var i = 0; i < rows.length; i++) {
      widgets.add(Row(
        children: [
          for (var j = 0; j < rows[i].length; j++) ...[
            Expanded(
              child: _optionButton('${rows[i][j]}', handicap == rows[i][j],
                  () => setState(() => handicap = rows[i][j])),
            ),
            if (j != rows[i].length - 1) const SizedBox(width: 8),
          ]
        ],
      ));
      if (i != rows.length - 1) widgets.add(const SizedBox(height: 8));
    }
    return widgets;
  }

  Widget _optionButton(String label, bool selected, VoidCallback onTap) =>
      ZenOptionButton(label: label, selected: selected, onTap: onTap);

  Future<void> _pickBot() async {
    final picked = await Navigator.of(context).push<BotProfile>(
      MaterialPageRoute(
        builder: (_) => BotPickerScreen(
          selected: selectedBot,
        ),
      ),
    );
    if (picked != null) {
      setState(() {
        selectedBot = picked;
        aiDifficulty = picked.engine;
      });
    }
  }
}

class _BotRow extends StatelessWidget {
  final BotProfile? bot;
  final VoidCallback onTap;
  const _BotRow({required this.bot, required this.onTap});

  @override
  Widget build(BuildContext context) {
    final scheme = Theme.of(context).colorScheme;
    final text = Theme.of(context).textTheme;
    final b = bot;
    return InkWell(
      onTap: onTap,
      borderRadius: BorderRadius.circular(14),
      child: Container(
        padding: const EdgeInsets.symmetric(horizontal: 12, vertical: 10),
        decoration: BoxDecoration(
          color: scheme.surfaceContainerHigh,
          borderRadius: BorderRadius.circular(14),
        ),
        child: Row(
          children: [
            Container(
              width: 40,
              height: 40,
              decoration: BoxDecoration(
                color: scheme.surface,
                shape: BoxShape.circle,
              ),
              alignment: Alignment.center,
              child: Text(b?.initials ?? '??',
                  style: text.labelLarge
                      ?.copyWith(fontWeight: FontWeight.w700)),
            ),
            const SizedBox(width: 12),
            Expanded(
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Text(b?.name ?? 'Pick an opponent',
                      style: text.bodyMedium
                          ?.copyWith(fontWeight: FontWeight.w600)),
                  if (b != null)
                    Text(
                      '${b.countryEmoji} · ${b.rankLabel} · ${b.style.label}',
                      style: text.labelSmall
                          ?.copyWith(color: scheme.onSurfaceVariant),
                    ),
                ],
              ),
            ),
            Icon(Icons.chevron_right, color: scheme.onSurfaceVariant),
          ],
        ),
      ),
    );
  }
}

class _SectionLabel extends StatelessWidget {
  final String text;
  const _SectionLabel(this.text);
  @override
  Widget build(BuildContext context) {
    final scheme = Theme.of(context).colorScheme;
    return Text(text,
        style: Theme.of(context)
            .textTheme
            .labelSmall
            ?.copyWith(color: scheme.onSurfaceVariant));
  }
}

class _RulesetSection extends StatelessWidget {
  final Ruleset selected;
  final ValueChanged<Ruleset> onSelect;
  const _RulesetSection({required this.selected, required this.onSelect});

  // Compact label for the chip (the canonical [Ruleset.label] is "New Zealand";
  // the chip uses "NZ" to fit the row).
  static String _chipLabel(Ruleset r) =>
      r == Ruleset.newZealand ? 'NZ' : r.label;

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        const _SectionLabel('RULESET'),
        const SizedBox(height: 8),
        SizedBox(
          height: 44,
          child: ListView.separated(
            scrollDirection: Axis.horizontal,
            itemCount: Ruleset.values.length,
            separatorBuilder: (_, __) => const SizedBox(width: 8),
            itemBuilder: (context, i) {
              final r = Ruleset.values[i];
              return SizedBox(
                width: r == Ruleset.trompTaylor ? 130 : 96,
                child: ZenOptionButton(
                  label: _chipLabel(r),
                  selected: r == selected,
                  onTap: () => onSelect(r),
                ),
              );
            },
          ),
        ),
      ],
    );
  }
}

class _ChipSection<T> extends StatelessWidget {
  final String title;
  final List<T> options;
  final T selected;
  final String Function(T) label;
  final ValueChanged<T> onSelect;

  const _ChipSection({
    super.key,
    required this.title,
    required this.options,
    required this.selected,
    required this.label,
    required this.onSelect,
  });

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        _SectionLabel(title),
        const SizedBox(height: 8),
        Row(
          children: [
            for (var i = 0; i < options.length; i++) ...[
              Expanded(
                child: ZenOptionButton(
                  label: label(options[i]),
                  selected: options[i] == selected,
                  onTap: () => onSelect(options[i]),
                ),
              ),
              if (i != options.length - 1) const SizedBox(width: 8),
            ],
          ],
        ),
      ],
    );
  }
}
